Prevent copying data where possible.

LumioseSil opened the pull request:
#### 5426d222c452763023a5543d95a8a79e537c1872
<pre>
Prevent copying data where possible.
<a href="https://bugs.webkit.org/show_bug.cgi?id=255164">https://bugs.webkit.org/show_bug.cgi?id=255164</a>

Reviewed by Darin Adler and Alex Christensen.

There is no reason to manually cast to CFStringRef
or create it from a string literal if there is a macro
that can do this for us.

Additionally, we can use CreateWithBytesNoCopy and other similar
functions when we know for the entire lifetime of the object, the
backing store is valid.

* Source/WTF/wtf/Assertions.cpp:
(WTF::createWithFormatAndArguments): Switch to
CFStringCreateWithCStringNoCopy.
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_newLineForElement): Switch to initWithCharactersNoCopy
(HTMLConverter::_addQuoteForElement): Ditto.
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
Deleted WebKit2HTTPProxyDefaultsKey and WebKit2HTTPSProxyDefaultsKey
refs.
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm: Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferences.mm:
(+[_WKSystemPreferences isCaptivePortalModeEnabled]): Switch to
CFStringCreateWithCStringNoCopy.
(+[_WKSystemPreferences setCaptivePortalModeEnabled:]): Ditto.
(+[_WKSystemPreferences setCaptivePortalModeIgnored:ignore:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferencesInternal.h: Make
LDMEnabledKey and
WKLockdownModeContainerConfigurationChangedNotification CFStrings.
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockNfcService.mm:
(-[WKMockNFTag initWithType:]): Switch to initWithBytesNoCopy.
(WebKit::MockNfcService::detectTags const): Ditto.
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::platformSetNetworkParameters): Define
WebKit2HTTPProxyDefaultsKey and WebKit2HTTPSProxyDefaultsKey
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextWidth.mm:
(TEST(WebKit, TextWidth)): Use CFString literal.

Canonical link: <a href="https://commits.webkit.org/282741@main">https://commits.webkit.org/282741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ee7c4a0ce99eeb15e690c230faa7791d504ebdd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68120 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14706 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66218 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51153 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14986 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51620 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10152 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67167 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55467 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32239 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36880 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12851 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13579 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/57210 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58840 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13180 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69819 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/63343 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12709 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58938 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8078 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55564 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59098 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6672 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85104 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9702 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39275 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15008 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41537 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40097 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/webauthn/WebAuthenticationConstants.h:
```diff
@@ -83,7 +83,9 @@ enum class ShouldZeroAAGUID : bool {
 
-constexpr const char LocalAuthenticatorAccessGroup[] = "com.apple.webkit.webauthn";
+#if defined(__OBJC__)
+static NSString* const LocalAuthenticatorAccessGroup = @"com.apple.webkit.webauthn";
```
This should only be a declaration in a header with a definition in a .mm file.

darinadler replied:
The use of `static` is wrong here

darinadler commented on Source/WTF/wtf/Assertions.cpp:
```diff
@@ -77,7 +77,7 @@ ALLOW_NONLITERAL_FORMAT_BEGIN
 #if USE(CF)
     if (strstr(format, "%@")) {
-        auto cfFormat = adoptCF(CFStringCreateWithCString(kCFAllocatorDefault, format, kCFStringEncodingUTF8));
+        auto cfFormat = adoptCF(CFStringCreateWithCStringNoCopy(kCFAllocatorDefault, format, kCFStringEncodingUTF8, kCFAllocatorNull));
```
Is it documented that `CFStringCreateWithFormatAndArguments` is guaranteed to not retain the format string? Or are we just guessing based on common sense?

AZero13 replied:
It's not guaranteed but it doesn't matter because format string isn't going anywhere.

darinadler replied:
I have no idea what distinction you are drawing. Saying “format string isn’t going anywhere” is saying you are guessing.

darinadler replied:
Here’s an example of how this could work in a surprising way: Imagine a version that cached the format string to optimize using the same format string multiple times and detecting if it’s the same. We need some guarantee that it doesn’t do something like that.

AZero13 replied:
> I have no idea what distinction you are drawing. Saying “format string isn’t going anywhere” is saying you are guessing.

I'm not guessing. Whatever format is pointing to is going to remain there for the entire lifetime of the CFString because it lasts until the end of the function.

That's a fact. It makes no sense for this to change between it being used as the backing store and when the CFString is released.

AZero13 replied:
There were in fact cases of this I considered and then immediately changed back because it was found that it was actually better for the string to be copied unconditionally.

darinadler commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1163,7 +1163,7 @@ static Color normalizedColor(Color color, bool ignoreDefaultColor, Element& elem
 {
     unichar c = NSLineSeparatorCharacter;
-    RetainPtr<NSString> string = adoptNS([[NSString alloc] initWithCharacters:&c length:1]);
+    RetainPtr<NSString> string = adoptNS([[NSString alloc] initWithCharactersNoCopy:&c length:1 freeWhenDone:NO]);
```
Is it documented that `-[NSString replaceCharactersInRange:withString:]` is guaranteed to never retain the passed in string, even in unusual cases like a one-character string? Or are we just guessing based on common sense?

AZero13 replied:
I actually heard that these one letter strings are handled differently in practice but that's not at all why I did this: I think that copying a one line string is not needed.

darinadler replied:
You mean one character string. I think that while possibly appealing, this is a bad idea for a change that will likely actually make performance a tiny bit worse.

AZero13 replied:
Honestly I'm surprised you mentioned this and not the fact it is a stack pointer, which I didn't know until an hour ago until someone else told me but yeah putting a stack pointer on an NSString, and in the end run you may be right that both calls will just become a tagged pointer, since these are single-char strings, so in practice this change likely won't matter. I should probably revert this if that is the case...

smfr replied:
I would like to see this addressed before landing.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferences.mm:
```diff
@@ -39,26 +39,25 @@ @implementation _WKSystemPreferences
 + (BOOL)isCaptivePortalModeEnabled
 {
-    auto key = adoptCF(CFStringCreateWithCString(kCFAllocatorDefault, WKLockdownModeEnabledKey.characters(), kCFStringEncodingUTF8));
+    auto key = adoptCF(CFStringCreateWithCStringNoCopy(kCFAllocatorDefault, WKLockdownModeEnabledKey.characters(), kCFStringEncodingUTF8, kCFAllocatorNull));
```
Is it documented that `CFPreferencesCopyValue` is guaranteed to never retain the passed in key? Or are we just guessing based on common sense?

AZero13 replied:
I mean characters is a const global variable of a C++ string.

Should not matter honestly because retain counts don't affect this.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferencesInternal.h:
```diff
@@ -26,7 +26,7 @@
 #import <wtf/text/ASCIILiteral.h>
 
-constexpr auto LDMEnabledKey = "LDMGlobalEnabled";
+const auto LDMEnabledKey = CFSTR("LDMGlobalEnabled");
```
I’m not sure that this is more efficient than the old idiom. Might be more elegant to not have the bridge cast, but I believe `@("")` is implemented entirely at compile time and `CFSTR` has some runtime cost. That’s also why we can’t use `constexpr` any more.

We should be cautious about changing coding style to something more elegant that might in fact be less efficient.

darinadler replied:
Also note that a bridge cast does not involve copying, so this is a part of the pull request that goes beyond the title, which refers to eliminating unnecessary copying.

darinadler reviewed (approved):
A reasonable cleanup. The benefit of these optimizations is tiny or negligible, so we should be cautious about taking any unnecessary risk.

smfr commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1357,7 +1357,7 @@ static Class _WebMessageDocumentClass()
 {
     unichar c = ((level % 2) == 0) ? (opening ? 0x201c : 0x201d) : (opening ? 0x2018 : 0x2019);
-    RetainPtr<NSString> string = adoptNS([[NSString alloc] initWithCharacters:&c length:1]);
+    RetainPtr<NSString> string = adoptNS([[NSString alloc] initWithCharactersNoCopy:&c length:1 freeWhenDone:NO]);
```
Same here; the string is created with a pointer to a stack variable. I would like to see this addressed before landing.

LumioseSil committed:
Prevent copying data where possible.
https://bugs.webkit.org/show_bug.cgi?id=255164

Reviewed by Darin Adler and Alex Christensen.

There is no reason to manually cast to CFStringRef
or create it from a string literal if there is a macro
that can do this for us.

Additionally, we can use CreateWithBytesNoCopy and other similar
functions when we know for the entire lifetime of the object, the
backing store is valid.

* Source/WTF/wtf/Assertions.cpp:
(WTF::createWithFormatAndArguments): Switch to
CFStringCreateWithCStringNoCopy.
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_newLineForElement): Switch to initWithCharactersNoCopy
(HTMLConverter::_addQuoteForElement): Ditto.
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
Deleted WebKit2HTTPProxyDefaultsKey and WebKit2HTTPSProxyDefaultsKey
refs.
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm: Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferences.mm:
(+[_WKSystemPreferences isCaptivePortalModeEnabled]): Switch to
CFStringCreateWithCStringNoCopy.
(+[_WKSystemPreferences setCaptivePortalModeEnabled:]): Ditto.
(+[_WKSystemPreferences setCaptivePortalModeIgnored:ignore:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferencesInternal.h: Make
LDMEnabledKey and
WKLockdownModeContainerConfigurationChangedNotification CFStrings.
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockNfcService.mm:
(-[WKMockNFTag initWithType:]): Switch to initWithBytesNoCopy.
(WebKit::MockNfcService::detectTags const): Ditto.
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::platformSetNetworkParameters): Define
WebKit2HTTPProxyDefaultsKey and WebKit2HTTPSProxyDefaultsKey
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextWidth.mm:
(TEST(WebKit, TextWidth)): Use CFString literal.

Canonical link: https://commits.webkit.org/282741@main
